### PR TITLE
Add guideline for version numbers

### PR DESCRIPTION
**Pull request by dreispt**
Rationale:

It is unfortunate that most modules don't include in the manifest the information of the major Odoo version they target. This is an important information since most of the time modules are incompatible with other Odoo version other than the one they target.
For example: it is frequent to find code in Github with an Odoo module in a "master" branch, but can't be sure of what Odoo version it targets by looking at it's manifest file.

This is so because in the past Odoo/Openerp automatically added the major version to module versions. Thus if your module was version "6.1.1", it would be displayed as "6.1.6.1.1".

It is no longer the case: [now](https://github.com/odoo/odoo/blob/8.0/openerp/modules/module.py#L368) Odoo does not do that if the major version is already appended to the module version!

Thus the Odoo target version can be included in the module version without being duplicated by the loading functions, and properly document the module's Odoo target version.

### CODE REVIEW

**Comment by dreispt** [comment]
@yvaucher what does "in progress" means? This is not a work in progress.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -17,6 +17,20 @@ section.
...
+The version number in the module manifest should be the Odoo major
+version (e.g. `8.0`) followed by the the module `x.y` version numbers.
+For example: `8.0.1.0` is expected fir the firest release of a 8.0
+module.
+
+The `x.y` version numbers follow the semantics `breaking.feature`:
+  * `x` increments when changes can break modules extending on it
+  * `y` increments when non-breaking new features are added
+
+Bugfixes are not required to increment module version numbers.
```
Comment (clonedagain) [inline CONTRIBUTING.md]: I'd add "unless a migration script is needed"

**Comment by yvaucher** [comment]
@dreispt in progress tag was added automatically by waffle.io I'll have a look to see if someone reactivated it for this project.

**Comment by clonedagain** [comment]
+1

**Comment by bwrsandman** [comment]
:+1:

**Comment by yvaucher** [comment]
AFAICT last time we talked about it version numbers was `x.y.z` not only `x.y`

This was proposed by @gurneyalex here:

https://lists.launchpad.net/openerp-community/msg02170.html
(maybe there is a more recent talk?)

Having a look at current version number in `__openerp__.py`, 

x.y (0.011 to 12.2): 730 (1055 with unported)
where (1x6.1, 21x 7.y, 4x 8.0 which could be following Odoo version)

x.y.z (0.0.1 to 7.0.1): 43 (89 with unported)
where (3x 7.0.z, 1x 6.1.0, 1x 5.0.2 which could be following Odoo version)

others:
1x 1
3
7.0.1.0
8.0.9.0.0
1.0r028
1.0r089
1.beta

Numbers tell us it's mainly `x.y` which is used. But 388 modules are version 1.0 and 156 are version 0.1 (of ported modules) and were certainly never bumped once. Otherwise quite some are in `x.y.z` format.

I would prefer not to narrow to much and keep the `z` for fixes. As `y` would be for added features.

By the way https://www.odoo.com/documentation/8.0/reference/module.html
links to http://semver.org/

**Comment by eLBati** [comment]
:+1:

**Comment by dreispt** [comment]
No need for a rushed merge, we should really get an an agreement on if it should be `x.y.z` or just `x.y`.
@yvaucher you statistics show preference for `x.y`. 

Personally, the current status is that version strings right now are mostly ignored, so I'm not sure if moving from that to having to bump `z` on each minor fix isn't a little too much. And with the major version prepended you get 5 (!) numbers in the version string.

IMO the Odoo major version plus `x` is a must have. 
`y` is nice to have. 
`z` is OK, maybe we could be optional?

Following the previous ML discussion, @gurneyalex opinion here would be nice.

**Comment by pedrobaeza** [comment]
I prefer only x.y. It's enough for the version number changes we made. I agree about using full version numbering (8.0.1.0)

**Comment by yvaucher** [comment]
In fact, 8.0 are 2 fixed version number so even if we write 8.0.x.y it doesn't give us more latitude.
And I'm starting to get used by x.y.z versioning. So at least could the `z` be indicated as optional?
I prefer to be able to bump an other number for fix only PR.

**Comment by pedrobaeza** [comment]
I don't understand why using 8.0.x.y incrementing y as needed (why not having 8.0.1.123?) is a problem for you.

**Comment by yvaucher** [comment]
For me fixes should not be numbered with minor change (added features).

Fixing bugs and adding features are quite no the same thing.

You should know that upgrading your modules from 1.0 to 1.0.5 gets you only fixes and going to 1.2 gives you new features.

**Comment by pedrobaeza** [comment]
But our experience tells us that nobody bump the version for fixing bugs. I only change the version number when migration scripts are needed.

**Comment by yvaucher** [comment]
However, fixed bug should in fact bump version as it will show in Odoo that a new version of the module is available. I think we should encouraging it even if we do it badly.

**Comment by pedrobaeza** [comment]
OK, let's add it to guidelines then and encourage this version bump in our reviews (I'll try to do my best about it, but it's difficult to remember). Maybe we can make any kind of automatic checking in Travis?

**Comment by dreispt** [comment]
Added version number for bugfixes.

**Comment by yvaucher** [comment]
@pedrobaeza I also admit I rarely do it, but I think it could improve user experience on our modules. Moreover I think it is important when you know you are fixing more than a typo as you want your patch to land ASAP.

@dreispt Thanks

:+1:

**Comment by gurneyalex** [comment]
I personnally like semantic versioning, but I won't push it in anyone's face. I'm in favor of including the target upstream version in the version number of our addons (but I don't feel like doing the mass commit script to achieve this. I'll run it if someone else writes it). So :+1:

The mailing list message refered to by @yvaucher was not met with much feedback at the time. 

(Any discussion about the versioning scheme for our addons should happen on the contributors mailing list, where it will get more visibility than on a github issue on a project watched by 37 persons.)

**Comment by lepistone** [comment]
IMO the only reason why we bump versions only for migrations is that at the moment there's no other use for them.

That is because until we keep having many modules per repo, we cannot tag the repo with the version, and we'll continue to use commit hashes as version numbers.

When that situation changes, version numbers will be git tags, and I'd really love to require releases for fixes, too. That way we'll be able to stop using git hashes.

**Comment by max3903** [comment]
Bumping version is not only for migration. It can be used also for packaging.

**Comment by yvaucher** [comment]
@lepistone Agreed on dividing the modules and rule them all. (Team and/or topics could be displayed on oca website to show the modules which are part of it. We could even have link module being part of the 2 topics it links which give us sometime a hard choice) As 9.0 will sooner or later be available it is an other talk we need to put under the spotlights (mailing list).

**Comment by max3903** [comment]
:+1: for x.y

Increment x when migration is required
Increment y when repackaging is required

**Comment by lepistone** [comment]
@max3903 what do you mean by packaging?

**Comment by nhomar** [comment]
Packaging is like "When data_data should be updated in a miron release.". did I understand correctly @max3903 ?

**Comment by max3903** [comment]
Packaging is about generating a .egg, .deb or .rpm file for a module. The package should be generated based on the module version (odoo-module_name-8.0.x.y.egg). If the module version does not change, then no new package are generated.

**Comment by lepistone** [comment]
thanks @max3903. I am not aware of a way to generate eggs of debs from odoo modules, but I'd be delighted to know!

**Comment by dreispt** [comment]
@lepistone I gave a stab at that and got [something](https://github.com/dreispt/pipo/blob/master/pipo.py) to work at the time.
@max3903 Don't you want to repackage when bugfixes are released? (assuming automated package generation)

**Comment by max3903** [comment]
@dreispt Yes, bump the version when you fix a bug, so a new package will be created.

**Comment by lepistone** [comment]
@dreispt @max3903 I knew about your pipo. I didn't know that anyone was actually packaging modules at this time.

**Comment by dreispt** [comment]
@lepistone AFAIK no one is, but I guess @max3903 has a secret plan...

**Comment by dreispt** [comment]
@max3903 ... then you want the bugfix `z` to be bumped, so that a package creation is triggered.

**Comment by max3903** [comment]
@dreispt no secret plan here, but at least it would be easier to do if sysadmin want to use their package manager to deploy Odoo modules.

**Comment by lepistone** [comment]
Thanks for the info. I think we'll all agree some kind of module packaging would be cool. So we can right now allow versioning for fixes, and when we somehow start to package or tag modules for real, we can discuss enforcing that.

**Comment by dreispt** [comment]
Can we merge this?

**Comment by pedrobaeza** [comment]
OK for me

**Comment by max3903** [comment]
I am still in favor of x.y and bumping y for a bugfix, but I can live with 3 numbers.

**Comment by nhomar** [comment]
:+1: for me..... I think we should merge this ASAP I can live with 3 or 2.

Just resuming what I read on my words to be sure I understand perfectly everything.

x.y.z

With x increase a migration script should come with the PR to the module.
with y increase a -u module -d database will be necesary.
with z a restart is enought.

Is this is Ok and I resumed correctly it is a :+1:  on your proposal

Even as a though we were facing this issues right now on our own repositories, but we do not use migration scripts yes that's why x.y was enought for us, but including migration on the roadmap it is fairly comprehensible.

And adding to this, IMHO inmediatly we merge this we must start to discuss the tagging for repositories, because between we review , merge and change the Versions we need stablish a tag to be have always a basement betwee what was before and what will be now. (in terms of module versions.)

Regards.

**Comment by pedrobaeza** [comment]
OK, merging then

**Comment by gurneyalex** [comment]
I wish this had been discussed on the contributors mailing list before merging. The exposition of these github discussions is quite low. Can we revert the commit and get a proper public debate on this?

**Comment by nbessi** [comment]
I agree with @gurneyalex please revert. That not a really "polite" way of doing things. This PR was merged before I can read it. As we are in period of holidays I'm probably not the only one. This change also means corrective works to be done and eventual side effect. IMHO that should be also discussed by OCA board.

**Comment by nhomar** [comment]
2015-08-05 1:36 GMT-05:00 Nicolas Bessi (nbessi) <notifications@github.com>:

> I agree with @gurneyalex <https://github.com/gurneyalex> please revert.
> That not a really "polite" way of doing things. This PR was merged before I
> can read it. As we are in period of holidays I'm probably not the only one.
> This change also means corrective works to be done and eventual side
> effect. IMHO that should be also discussed by OCA board.
>

It had more than 2 approvals and it was clear that the ponts need more work.

I can revert it (it is possible also).

but IMHO it stay 10 days open isn't that a good period of time?


-- 
--------------------
Saludos Cordiales

CEO at Vauxoo <https://www.vauxoo.com> Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
<http://about.me/nhomar?promo=email_sig>

**Comment by nhomar** [comment]
Sorry I didn't read the @pedrobaeza mention.... I will let him revert it.

BTW, can you tell here if you are not agreed?

**Comment by pedrobaeza** [comment]
Well, I think that now it's better to open the discussion in the mailing list with this pre-agreement taken, and if any change, we amend the text. Don't you think?

**Comment by nhomar** [comment]
2015-08-05 3:22 GMT-05:00 Pedro M. Baeza <notifications@github.com>:

> Well, I think that now it's better to open the discussion in the mailing
> list with this pre-agreement taken, and if any change, we amend the text.
> Don't you think?


I am open to both option.

Left merged and open the discusion again and revert the commit but I think
Alex and Nicolas have a point Pedro.

If 2 persons approve the merge is done, but if afterwards and inside a
logic gap of time 2 complains it must be reverted, then in this case I
think it must be reverted and reopen the discussion on the mailing list.

(even if I am not agreed that's the meaning of democracy )




-- 
--------------------
Saludos Cordiales

CEO at Vauxoo <https://www.vauxoo.com> Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
<http://about.me/nhomar?promo=email_sig>

**Comment by pedrobaeza** [comment]
OK, no problem, reverting...

**Pull request merged**